### PR TITLE
sql: introduce SHOW CREATE TABLE statement

### PR DESCRIPTION
**Pull request by ImeevMA**
This patch introduces the SHOW CREATE TABLE statement. This statement can be used to obtain a description of a space in the form of a corresponding CREATE TABLE statement.

Part of #8098

@TarantoolBot document
Title: SHOW CREATE TABLE statement

Statement can be used to obtain a description of a space in the form of a corresponding CREATE TABLE statement.

There is two syntax for SHOW CREATE TABLE statement: 1) Get description of single space:
```
SHOW CREATE TABLE table_name;
```

2) Get descriptions of all available non-system spaces:
```
SHOW CREATE TABLE;
```

### CODE REVIEW

File: src/box/sql/show.c
Comment (Gerold103) [inline src/box/sql/show.c:142]: I would say this behaviour is quite unexpected:
```
tarantool> box.execute('SHOW CREATE TABLE "404"')
---
- metadata:
  - name: CREATE TABLE statement
    type: string
  rows: []
...
```
```
tarantool> s = box.schema.create_space('test')
---
...

tarantool> box.execute('SHOW CREATE TABLE "test"')
---
- metadata:
  - name: CREATE TABLE statement
    type: string
  rows:
  - ['/* Space "test" cannot be defined in this statement. */']
...
```
Why aren't these normal errors? Users have no sane way to see if the statement has failed. They would need to check in their code for no rows and for `/*` in the beginning of the statement?

File: src/box/sql/build.c
Comment (Gerold103) [inline src/box/sql/build.c:3442]: Why do you need so many additional opcodes if you anyway added `OP_ShowTable`? Couldn't space name simply be its direct argument?

**Review by Gerold103 — CHANGES_REQUESTED** [review]
Thanks for the patch!

File: src/box/sql/show.c
Comment (ImeevMA) [reply src/box/sql/show.c:131]: This case is checked in the test on line 154. This is when the outer space does not have a field referenced by the local space. This situation is possible in Lua, because the core FK checks the availability of a field at run time, not at creation.

File: src/box/sql/show.c
Comment (ImeevMA) [reply src/box/sql/show.c:2]: I add only one `\n`, I think the second one added when the result is shown in console:
```
tarantool> box.execute([[create table t(i int primary key, a int);]])
---
- row_count: 1
...

tarantool> print(box.execute([[show create table t;]]).rows[1][1])
CREATE TABLE "T"(
"I" INTEGER NOT NULL,
"A" INTEGER,
CONSTRAINT "pk_unnamed_T_1" PRIMARY KEY("I"));
---
...
tarantool> box.execute([[show create table t;]]).rows[1][1]
---
- 'CREATE TABLE "T"(

  "I" INTEGER NOT NULL,

  "A" INTEGER,

  CONSTRAINT "pk_unnamed_T_1" PRIMARY KEY("I"));'
...
```

File: src/box/sql/show.c
Comment (ImeevMA) [reply src/box/sql/show.c:142]: I don't think the statement failed in the second case. As for the first one, since we didn't find anything, it seems logical to me that we return empty result. Either way, let's ask @unera.

File: src/box/sql/build.c
Comment (Gerold103) [reply src/box/sql/build.c:3442]: Ah, I see, it is called for each tuple during iteration. Not once per vdbe exec.

File: src/box/sql/show.c
Comment (Gerold103) [reply src/box/sql/show.c:142]: But this is not a `"CREATE TABLE"` statement. So the request couldn't return what was expected. That is an error. You even literally call this message in your code `/** Buffer for errors. */`.

File: changelogs/unreleased/gh-8098-show-create-table.md
Comment (ImeevMA) [reply changelogs/unreleased/gh-8098-show-create-table.md]: True, fixed!

File: src/box/sql/build.c
Comment (ImeevMA) [reply src/box/sql/build.c]: Thanks, replaced by "statement".

File: src/box/sql/show.c
Comment (ImeevMA) [reply src/box/sql/show.c]: True, reworked using `struct StrAccum` and `sqlStrAccum*()` functions.

File: src/box/sql/show.c
Comment (ImeevMA) [reply src/box/sql/show.c:1]: Thanks for noticing and suggesting! Fixed.

File: src/box/sql/show.c
Comment (ImeevMA) [reply src/box/sql/show.c:106]: Replaced error messages by actual errors. However, this actually affects `SHOW CREATE TABLE;` statement, since one wrong space and no information will be given. Maybe we allow error messages for all cases except `Space 'space_name' does not exist`?

Replaced error messages with real errors. However, this actually affects the `SHOW CREATE TABLE;` statement, as one wrong space and no proper result will be provided. Maybe we allow error messages for all cases except `Space 'space_name' does not exist`?

Comment (unera) [reply src/box/sql/show.c:106]: I agree, unknown name should throw an error.

Comment (unera) [reply src/box/sql/show.c:106]: update: we should generate an error only if `SHOW CREATE TABLE t` links to unknown (or inaccessible) `t`.

The operator can contain some comments:
- about indexes, that can not be created by SQL
- about any the other problems

In an extreme case, this statement will only contain comments

File: src/box/sql/show.c
Comment (Gerold103) [inline src/box/sql/show.c:35]: To be on the safe side I suggest to use here `CFORMAT` attribute (from `trivia/util.h`). It will validate the format string and args at compile time.

File: src/box/sql/show.c
Comment (Gerold103) [inline src/box/sql/show.c:230]: We already have `strtolower()` / `strtolowerdup()` in `trivia/util.h`. Maybe worth adding `strtoupper()` / `strtoupperdup()`.

File: src/box/sql/show.c
Comment (Gerold103) [inline src/box/sql/show.c:280]: There are quite many of these triplets - name new + append + xfree. I would suggest to introduce `sql_desc_append_escape(desc, name)` which would the the same. It would force you to write sometimes not much less code, but you won't have to keep track of all those xfree:
```C
sql_desc_append(desc, ",\nCONSTRAINT ");
sql_desc_append_escape(desc, pk->def->name);
sql_desc_append(desc, "PRIMARY KEY(");
```
In some places it will turn into just 1 or 2 lines. Up to you.

File: src/box/sql/show.c
Comment (Gerold103) [inline src/box/sql/show.c:389]: `space_is_memtx()` and `space_is_vinyl()` from `space.h`.

File: test/sql-luatest/show_create_table_test.lua
Comment (Gerold103) [inline test/sql-luatest/show_create_table_test.lua:62]: You might want to consider this simplification (you do gsub ~35 times in this file right now):

```Diff
@@ -14,14 +14,26 @@ end)
 
 g.test_show_create_table_one = function()
     g.server:exec(function()
-        local _, err = box.execute([[SHOW CREATE TABLE t;]])
+        rawset(_G, 'test_exec', function(...)
+            local res = box.execute()
+            t.assert(res ~= nil)
+            for i = 1, #res.rows do
+                local field = res.rows[i][1]
+                if field ~= nil then
+                    res.rows[i]:update({{'=', 1, field:gsub('\n', '')}})
+                end
+            end
+            return res
+        end)
+
+        local _, err =  box.execute([[SHOW CREATE TABLE t;]])
         t.assert_equals(err.message, [[Space 'T' does not exist]])
 
         box.execute([[CREATE TABLE t(i INT PRIMARY KEY, a INT);]])
-        local ret = box.execute([[SHOW CREATE TABLE t;]])
+        local ret = _G.test_exec([[SHOW CREATE TABLE t;]])
         local res = [[CREATE TABLE "T"("I" INTEGER NOT NULL,"A" INTEGER,]]..
                     [[CONSTRAINT "pk_unnamed_T_1" PRIMARY KEY("I"));]]
-        t.assert_equals(ret.rows[1][1]:gsub('\n', ''), res)
+        t.assert_equals(ret.rows[1][1], res)
         box.execute([[DROP TABLE t;]])
 
         local sql = [[CREATE TABLE "T"("I" INTEGER NOT NULL,]]..
@@ -35,8 +47,8 @@ g.test_show_create_table_one = function()
                     [[CONSTRAINT "C5" FOREIGN KEY("I", "A") ]]..
                     [[REFERENCES "T"("A", "C")) WITH ENGINE = 'vinyl';]]
         box.execute(sql)
-        ret = box.execute([[SHOW CREATE TABLE t;]])
-        t.assert_equals(ret.rows[1][1]:gsub('\n', ''), sql)
+        ret = _G.test_exec([[SHOW CREATE TABLE t;]])
+        t.assert_equals(ret.rows[1][1], sql)
         box.execute([[DROP TABLE t;]])
     end)
 end
@@ -49,7 +61,7 @@ g.test_show_create_table_all = function()
         box.execute([[CREATE TABLE t1(i INT PRIMARY KEY, a INT);]])
         box.execute([[CREATE TABLE t2(i INT PRIMARY KEY, a INT);]])
         box.execute([[CREATE TABLE t3(i INT PRIMARY KEY, a INT);]])
-        local ret = box.execute([[SHOW CREATE TABLE;]])
+        local ret = _G.test_exec([[SHOW CREATE TABLE;]])
         local res1 = [[CREATE TABLE "T1"("I" INTEGER NOT NULL,"A" INTEGER,]]..
                      [[CONSTRAINT "pk_unnamed_T1_1" PRIMARY KEY("I"));]]
         local res2 = [[CREATE TABLE "T2"("I" INTEGER NOT NULL,"A" INTEGER,]]..
@@ -57,9 +69,9 @@ g.test_show_create_table_all = function()
         local res3 = [[CREATE TABLE "T3"("I" INTEGER NOT NULL,"A" INTEGER,]]..
                      [[CONSTRAINT "pk_unnamed_T3_1" PRIMARY KEY("I"));]]
         t.assert_equals(#ret.rows, 3)
-        t.assert_equals(ret.rows[1][1]:gsub('\n', ''), res1)
-        t.assert_equals(ret.rows[2][1]:gsub('\n', ''), res2)
-        t.assert_equals(ret.rows[3][1]:gsub('\n', ''), res3)
+        t.assert_equals(ret.rows[1][1], res1)
+        t.assert_equals(ret.rows[2][1], res2)
+        t.assert_equals(ret.rows[3][1], res3)
```
and etc similarly.

File: test/sql-luatest/show_create_table_test.lua
Comment (Gerold103) [inline test/sql-luatest/show_create_table_test.lua:166]: AFAIU, the purpose of `SHOW CREATE TABLE` usually is to export the schema into an SQL script and then re-apply it anew in some other place. And I still see problems with it in the current API:

* If you silently ignore non-printable parts like unknown foreign keys, then you generate a valid but just wrong SQL statement. People will print it into a script, apply, won't see any errors, and then stuff will explode in their tests. Not at the time of the SQL script generation. This is sad. It makes the feature hardly usable as is, I think.

* If a space couldn't be serialized into SQL, as I can see, you still return it as an error-comment. This is fixed for the individual `SHOW CREATE TABLE <name>`, but not for showing all tables. Again, the scenario about schema export will break here. People will need to write boilerplate code around this to check for "statements" which start from `/*` and thus aren't really statements.

I understand the problems though. You want to return the error somehow. So I would suggest to either fail if anything couldn't be serialized, or just skip such spaces from the result (sad, no full schema, but the returned spaces will be at least full and valid), or return 2 fields: `statement` and `error` or `message`. `statement` is always fully serialized space-SQL or `nil`. `error` / `message` is either `nil` or the message. Then people at least will be able to check for individual errors normally, instead of doing `res.rows[i]:starts('/*')` which is just extremely weird. Note, that even this hack wouldn't help against the code block which I commented. When you appended the error somewhere in the end.

You summoned your product-pokemon last time? ;) Now lets do a battle - I summon mine! @Mons, ATTACK! (please)

**Review by Gerold103 — COMMENTED** [review]
Thanks for the fixes!

File: src/box/sql/show.c
Comment (Gerold103) [reply src/box/sql/show.c:88]: But you are not **doing** the create-table here, are you? You just **generate** the statement. You don't execute it. So what is the problem anyway? You have the name - you can use it. You check it now in 3 places (which I would recommend to extract into a function if you intend to keep them even after this comment), but never use the resulting fieldno. Because the field name was enough. Why care in this `SHOW` statement if the name matches with any fieldno if you don't use it? If this field will be missing at the time of the resulting create-table execution (if it will be executed at all), then just let it fail there.

I can expand my reasoning - *"why try to guess what CREATE TABLE fails at, if you are not doing it here?"*. You only need to serialize the definition precisely. Fail only when it is not possible (like field type mismatch in index and space). This concerns not only FK fieldnos. There might be other restrictions which CREATE TABLE has and you do not account them here. For instance, we have a max column count limit - `SQL_MAX_COLUMN`. You don't check it here, but the resulting `box.execute(res.rows[1][1])` will fail if you serialized a non-SQL space with more columns.

Up to you. Those checks just seem as harmless as useless.

File: test/sql-luatest/show_create_table_test.lua
Comment (unera) [reply test/sql-luatest/show_create_table_test.lua:166]: > If you silently ignore non-printable parts like unknown foreign keys, then you generate a valid but just wrong SQL statement.

some things is not provided by `CREATE TABLE` operator by design. For example non-unique indexes, etc.

So the pattern will never work in ALL user scenarios. I think it is a normal situation.
Worse if a user asks the list and receive an error because one item of the list contains non-printable index.

Comment (Mons) [reply test/sql-luatest/show_create_table_test.lua:166]: I mostly agree with @Gerold103. Explicit error in such a case is much better than silent hidden rake of wrong SQL.
Nevertheless I'd like to have partially dumped statement for such "undumpable" spaces, rather than none at all.
So, I'd expect explicit error in case of some things, that could not be dumped. And additional modifier, that will dump a space anyway, with errors shown as comments (like `show create table <identifier> ignore erorrs`)

Comment (ImeevMA) [reply test/sql-luatest/show_create_table_test.lua:166]: I discussed with @unera and we decided to introduce 4 versions of `SHOW CREATE TABLE`:
1) `SHOW CREATE TABLE <table_name>;` - return a description of the space. On any problem, it gives an error;
2) `SHOW CREATE TABLE <table_name> INCLUDING ERRORS;` - return a description of the space. If the table does not exist, an error is thrown, any other problems are included as comments to `CREATE TABLE`;
3) `SHOW CREATE TABLE;` - return descriptions of all non-system available spaces that can be described with `SHOW CREATE TABLE <table_name>;`. Spaces that cannot be described with `SHOW CREATE TABLE <table_name>;` are ignored;
4) `SHOW CREATE TABLE INCLUDING ERRORS;` -  return descriptions of all non-system available spaces. Any problems are included as comments to `CREATE TABLE`.

There is one more important change: if the INCLUDING ERRORS clause is used, the description will be returned even if the description is not working `CREATE TABLE`. For example, for a space created with `box.schema.space.create('a')`, the description would be:
```
CREATE TABLE "a";
/* Space "a" cannot be defined in this statement: PRIMARY KEY missing */
/* Space "a" cannot be defined in this statement: SQL does not support space without format */
```

Comment (unera) [reply test/sql-luatest/show_create_table_test.lua:166]: >
```sql
CREATE TABLE "a";
/* Space "a" cannot be defined in this statement: PRIMARY KEY missing */
/* Space "a" cannot be defined in this statement: SQL does not support space without format */
```

Could we place comments into brackets?

```sql
CREATE TABLE "a" (
/* Space "a" cannot be defined in this statement: PRIMARY KEY missing */
/* Space "a" cannot be defined in this statement: SQL does not support space without format */
);
```

Comment (unera) [reply test/sql-luatest/show_create_table_test.lua:166]: If a table is created, how it can throw an error 'Wrong engine'?

Anyway, we can place the comment inside brackets (As the first or last comment).

Comment (ImeevMA) [reply test/sql-luatest/show_create_table_test.lua:166]: For example:
```
SHOW CREATE TABLE "_space";
```

Comment (Gerold103) [reply test/sql-luatest/show_create_table_test.lua:166]: Why so complicated? Did you all read my comment fully? Citing, I wrote:

"... *or return 2 fields: `statement` and `error` or `message`. `statement` is always a fully serialized space-SQL or `nil`. `error` / `message` is either `nil` or an error message. Then people at least will be able to check for individual errors normally* ...".

What is wrong with this? You get complete error messages + you don't fail the whole `SHOW CREATE TABLE` because of just one space + you don't have incorrect SQL statements. And you don't need to introduce multiple versions of `SHOW` with weird syntax.

Comment (Mons) [reply test/sql-luatest/show_create_table_test.lua:166]: > Why so complicated?

I think, because in case of minor errors I'd like to receive my SQL statement with some comments, not nil and error message and then write this statement by hands. There are often no ability to fix an issue to make dump successful.

File: src/box/sql/show.c
Comment (ImeevMA) [reply src/box/sql/show.c]: Removed this check.

File: src/box/sql/show.c
Comment (ImeevMA) [reply src/box/sql/show.c:69]: Thanks, added.

File: src/box/sql/show.c
Comment (ImeevMA) [reply src/box/sql/show.c]: Thanks! Done.

File: src/box/sql/util.c
Comment (ImeevMA) [reply src/box/sql/util.c:1256]: Done. If the first character is a letter, the name is equal to its normalized version, and the escaped version is exactly 2 bytes longer, then I use the normalized name.

File: test/sql-luatest/show_create_table_test.lua
Comment (ImeevMA) [reply test/sql-luatest/show_create_table_test.lua]: Instead, I decided to add '\n' in expected result. However, I still added two helper functions.

File: test/sql-luatest/show_create_table_test.lua
Comment (ImeevMA) [reply test/sql-luatest/show_create_table_test.lua]: I'm not sure if we should return both a not-NULL result and an error in the same answer. Also, there is a use case similar to this case in the downgrade feature.

**Review by Gerold103 — COMMENTED** [review]
Thanks for the fixes!

File: src/box/sql/show.c
Comment (ImeevMA) [reply src/box/sql/show.c]: Thank you for noticing! Fixed.

File: src/box/sql/show.c
Comment (ImeevMA) [reply src/box/sql/show.c:241]: No, this is actually different. In ANSI query
```
CREATE TABLE t(i INT PRIMARY KEY, a INT CHECK(i > 10));
```
is the same as
```
CREATE TABLE t(i INT PRIMARY KEY, a INT, CHECK(i > 10));
```
since all column constraints are just another syntax for table constraints. This is not so for us. For example, we can set constraint "i > 10" in Lua for field `a` and it will check field `a`, since no names are checked in core field constraints. However, if we allow `SHOW CREATE TABLE` for such cases we receive
```
CREATE TABLE t(i INT PRIMARY KEY, a INT CHECK(i > 10));
```
which could not be executed from out standpoint and will have another result for ANSI standpoint.

If only we were able to avoid so-called "field constraints", we wouldn't have such problems. But we seek for speed here, it seems.

File: src/box/sql/show.c
Comment (ImeevMA) [reply src/box/sql/show.c:405]: Originally this statement was inspired by `SHOW CREATE TABLE` from MySQL. This statement in MySQL actually ignores indexes that cannot be described in `CREATE TABLE`.

There is also `.schema` and `.schema table_name` which return all `CREATE TABLE` and `CREATE INDEX` for all spaces and spaces with the given name, respectively.

I suggested doing something that works like `.schema table_name`, but `SHOW CREATE TABLE` seems more appealing to PM.

File: src/box/sql/show.c
Comment (Gerold103) [reply src/box/sql/show.c:241]: So like I said, it can't be executed anyway. But you still care to validate it. Which means this is exactly like the situation with fieldno vs name was, and exactly like the tuple constraints which currently are not validated and won't be executed either (I gave you the example above). I don't see why keep this particular validation, if others are not present or don't help anyway, just like this one.

Comment (ImeevMA) [reply src/box/sql/show.c:241]: I thought we decided that if the result was returned by `SHOW CREATE TABLE tablename;`, then that means the result should be executable. If I skip the check, this rule may be violated. Not sure if this is correct.

Comment (Gerold103) [reply src/box/sql/show.c:241]: This is already violated. I gave you an example above. Can copy-paste it again:
```Lua
box.cfg{}
body = 'A > 100 AND A < 200'
box.schema.func.create('check', {is_deterministic = true, body = body, language = 'SQL_EXPR'})
s = box.schema.create_space('test', {
    format = {{'ID', 'unsigned'}},
    constraint = {check = 'check'}})
_ = s:create_index('pk')
box.execute('SHOW CREATE TABLE "test"')
```
```
---
- metadata:
  - name: statement
    type: string
  rows:
  - ['CREATE TABLE "test"(

      ID UNSIGNED NOT NULL,

      CONSTRAINT "pk" PRIMARY KEY(ID),

      CONSTRAINT "check" CHECK(A > 100 AND A < 200));']
...
```
The returned SQL statement can't be executed. It will throw an error. That means the `SHOW CREATE TABLE` returned a bad result.

**Review by Gerold103 — APPROVED** [review]
Thanks for the fixes! I still see problems here which make the feature barely usable except manually in an interactive console (as an alternative to looking directly at `_space` content). But I don't want to argue, so ack.

File: src/box/sql/show.c
Comment (ImeevMA) [reply src/box/sql/show.c:129]: No, this is for the case when an error "wrong space engine" is printed. We still have to end statement with `);`.

File: src/box/sql/show.c
Comment (ImeevMA) [reply src/box/sql/show.c:241]: I will fill a bug. I am not sure which of these are more correct.

**Review by Gerold103 — APPROVED** [review]
Thanks for the patch! One comment, then need a second review.

File: changelogs/unreleased/gh-8098-show-create-table.md
Comment (p7nov) [inline changelogs/unreleased/gh-8098-show-create-table.md:3]:
```suggestion
* Introduced the `SHOW CREATE TABLE` statement (gh-8098).
```

**Review by p7nov — APPROVED** [review]

File: src/box/sql/show.c
Comment (ImeevMA) [reply src/box/sql/show.c:104]: Thank you for the review! Fixed. This is not necessary as `acc` is only populated after all the checks have passed, but there is no proper way to check that `acc` is empty, so adding `sqlStrAccumReset()` seems to be the only correct way to make sure nothing is leaked.

File: changelogs/unreleased/gh-8098-show-create-table.md
Comment (ImeevMA) [reply changelogs/unreleased/gh-8098-show-create-table.md]: Thanks! Fixed.

**Review by p7nov — APPROVED** [review]

**Review by locker — APPROVED** [review]

**Comment by locker** [comment]
@ImeevMA There are compilation errors on some platforms. Please fix.

**Comment by ImeevMA** [comment]
@locker fixed!
<details>
<summary>Diff</summary>

```DIff
diff --git a/src/box/sql/show.c b/src/box/sql/show.c
index 4d2d0a4a4..668c2f1a1 100644
--- a/src/box/sql/show.c
+++ b/src/box/sql/show.c
@@ -7,14 +7,10 @@
 
 #include "sqlInt.h"
 #include "mem.h"
-#include "schema.h"
-#include "sequence.h"
-#include "coll_id_cache.h"
-#include "tuple_constraint_def.h"
-
-enum {
-       TYPE_STR_SIZE = 16,
-};
+#include "box/schema.h"
+#include "box/sequence.h"
+#include "box/coll_id_cache.h"
+#include "box/tuple_constraint_def.h"
 
 /** An objected used to accumulate a statement. */
 struct sql_desc {
```

</details>

**Comment by ImeevMA** [comment]
@locker sorry, I found out that I wrote `Part of` instead of `Closes`. Will fix.

**Pull request merged**
